### ISSUE TITLE
Require semicolon or newline before method body

### ISSUE THREAD

**Issue by straight-shoota** [crystal-lang/crystal#13386]
The compiler currently does not require a clear delimiter between the head and body of a method.

For example, `def foo : Int32 1 end` is valid syntax.

There are some oddities: `def foo 1 end` is not valid, but `def foo() 1 end` is.

Some typos can cause surprising behaviour. The following code is valid:
```cr
def foo : Array (String)
  [1]
end
```

The return type restriction is `Array`. `(String)` is parsed as the first expression in the method body.

A variation of that appears in #6191:
```cr
def foo(x : String) : Tuple (String, String) # Syntax error: unterminated parenthesized expression
  {x, x}
end
```

I think requiring an explicit delimiter could improve the last examples because they would lead to (meaningful) syntax errors.

This comment form @FnControlOption in https://github.com/crystal-lang/crystal/pull/11854#issuecomment-1518168910 makes a concrete suggestion:

> `def f(x) puts x end` is both valid Ruby and Crystal, and requiring a semicolon in this case seems unnecessarily restrictive. I think a semicolon/newline should only be required if the method header (a) has both no parameters and no parentheses; or (b) has anything after the parentheses (i.e. an explicit return type and/or any generic type parameters).

I'm not sure about (a). A `def` without parenthesis cannot have parameters, so this case should be pretty unambiguous.

**Comment by straight-shoota** [comment]
Yeah, it's probably better to be strict about parameterless defs, even though it's technically unambiguous in Crystal.

**Comment by straight-shoota** [comment]
Actually, the formatter already takes care of removing the potential ambiguous syntax.

The first example from the OP formats like this:
```cr
def foo : Array
  (String)
  [1]
end
```

So the alternative is definitely considered ill-formed. But it's actually unambiguous and maybe it's fine to keep accepting it as valid code?
If you want your code to be formatted in a sane way, you should use the formatter anyway =)

**Comment by stefandd** [comment]
I believe this is related!

```Crystal
def foo : Int32
	switch = true
    if switch 0 else 0 end # Syntax error: Error: unexpected token: "else"
end
puts foo
```
The above doesn't compile but with an inserted ';', it prints `10`.

```Crystal
def foo : Int32
	switch = true
    if switch ; 0 else 0 end # works
end
puts foo
```

**Comment by ysbaddaden** [comment]
@stefandd `switch 10` is interpreted as a function call with args, and `else` appears out of nowhere (syntax error). Adding the semicolon tells crystal to _not_ look for args, and it parses correctly.
